### PR TITLE
Add unchecked warnings failures for src

### PR DESCRIPTION
**Pull request by jimmyjames**
### Changes

_This is primarily a tech debt and future preparedness PR_

`APIException` currently performs an unsafe (in compiler terms) cast of the error description to a `Map<String, Object>`. Functionally, this shouldn't be an issue since a `ClassCastException` would only come from a change in the error schema, which shouldn't happen.

However, it does cause an unchecked cast warning, which we should typically not have or suppress as needed. This change does the following:
- Updates `APIException` to only create a `PasswordStrengthErrorParser` if the error description is a map (currently it would fail with a runtime exception if not a string)
- Suppresses the unchecked exception when casting the error description to `Map<String, Object>`
- Adds tests to cover this change and the `obtainExceptionMessage` behavior in general
- Moves the compiler options to `compileJava` and `compileTestJava`
- Adds the `-Xlint:unchecked` compiler flag for src (not for tests)

### Testing

Please describe how this can be tested by reviewers. Be specific about anything not tested and reasons why. If this library has unit and/or integration testing, tests should be added for new functionality and existing tests should complete without errors. 

- [X] This change adds test coverage
- [ ] This change has been tested on the latest version of the platform/language or why not

### Checklist

- [X] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [X] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [X] All existing and new tests complete without errors

### CODE REVIEW

File: src/main/java/com/auth0/exception/APIException.java [hunk 1]
```diff
@@ -136,7 +136,8 @@ private static String obtainExceptionMessage(Map<String, Object> values) {
             Object description = values.get("description");
             if (description instanceof String) {
                 return (String) description;
-            } else {
+            } else if (description instanceof Map){
+                @SuppressWarnings("unchecked")
                 PasswordStrengthErrorParser policy = new PasswordStrengthErrorParser((Map<String, Object>) description);
                 return policy.getDescription();
             }
```
Comment (lbalmaceda) [inline src/main/java/com/auth0/exception/APIException.java:143]: else...? 
there should always be one default case. Maybe we could set a default message/description value here?

File: src/test/java/com/auth0/exception/APIExceptionTest.java [hunk 2]
```diff
@@ -13,16 +14,85 @@
...
         values = new HashMap<>();
     }
 
+    @Test
+    public void shouldGetMessageFromErrorDescription() {
+        values.put("error_description", "some error description");
+        values.put("description", "some description");
+        values.put("message", "some message");
+        values.put("error", "some error");
+        APIException apiException = new APIException(values, ERROR_CODE);
```
Comment (lbalmaceda) [inline src/test/java/com/auth0/exception/APIExceptionTest.java:31]: the ERROR_CODE value is never checked later. Maybe add one test for asserting it gets set somewhere?

**Review by lbalmaceda — CHANGES_REQUESTED** [review]
two small comments

File: src/main/java/com/auth0/exception/APIException.java (hunk 1 above)
Comment (jimmyjames) [reply src/main/java/com/auth0/exception/APIException.java:143]: I think that makes sense. I'll update with a default if the description is not a string or a map.

File: src/test/java/com/auth0/exception/APIExceptionTest.java (hunk 2 above)
Comment (jimmyjames) [reply src/test/java/com/auth0/exception/APIExceptionTest.java:31]: it is checked, through the `EXPECTED_ERROR_MESSAGE_PREFIX` constant above.

File: src/main/java/com/auth0/exception/APIException.java (hunk 1 above)
Comment (jimmyjames) [reply src/main/java/com/auth0/exception/APIException.java:143]: Actually, I did add a test for this very case (where the description is not a string or a map):

https://github.com/auth0/auth0-java/blob/eb1288f4370fb6651eacfc559ba788eeb2afe43e/src/test/java/com/auth0/exception/APIExceptionTest.java#L63

It verifies that in that case (which really should never happen, and as of today would just cause a runtime class cast exception) the error handling falls through the rest of the condition checks and returns "Unknown exception". I think that's the correct behavior; if the description is not one of the known types (map or string), we just continue to check for other values to return (message and error), before returning "Unknown exception".

File: src/test/java/com/auth0/exception/APIExceptionTest.java (hunk 2 above)
Comment (lbalmaceda) [reply src/test/java/com/auth0/exception/APIExceptionTest.java:31]: Correct. I thought there was a getter for the error code as well.

**Review by lbalmaceda — APPROVED** [review]

**Pull request merged**